File: GalleryAPI/galleryapi.js
// Create express app
var express = require("express")
var app = express()
var md5 = require("md5")
var db = require("./dbhandler")

var bodyParser = require("body-parser");
app.use(bodyParser.urlencoded({ extended: false }));
app.use(bodyParser.json());

// Server port
var HTTP_PORT = 3232
// Start server
app.listen(HTTP_PORT, () => {
    console.log("Server running on port %PORT%".replace("%PORT%",HTTP_PORT))
});

// Endpoints
app.get("/", (req, res, next) => {
    res.json({"message":"Ok"})
    console.log("enter to web.")
});

// Get data quary.
app.get("/api/:field/:table/:condition", (req, res, next) => {
    var field =  req.params.field, table = req.params.table, condition = req.params.condition;

    if (condition == "null")
        condition = " ";

    var sql = "SELECT " + field + " FROM " + table + " " + condition;
    var params = []

    db.all(sql, params, (err, rows) => {
        if (err) // If there is an error.
        {
            console.log("Error select quary")
            res.status(400).json({"error":err.message});
            return;
        }
        // Else return the user data.
        console.log("Data was return. (select)")
        res.json({
            "message":"success",
            "data":rows
        })
    });
});

// Add new user.
app.post("/api/user/", (req, res, next) => {
    var errors=[]
    if (!req.body.password){errors.push("No password specified");}
    if (errors.length)
    {
        res.status(400).json({"error":errors.join(",")});
        console.log("Error adding user.")
        return;
    }

    var data = {
        name: req.body.name,
        password: md5(req.body.password)
    }

    var sql ='INSERT INTO Users (Name, Password) VALUES (?,?)'
    var params =[data.name, data.password]
    
    db.run(sql, params, function (err, result) {
        if (err)
        {
            res.status(400).json({"error": err.message})
            console.log("Error adding user.")
            return;
        }
        console.log("User was added.")
        res.json({
            "message": "success",
            "data": data,
            "id" : this.lastID
        })
    });
});

// Add new album
app.post("/api/album/", (req, res, next) => {
    var errors=[]
    if (!req.body.user_id){errors.push("No User_id specified");}
    if (!req.body.creation_date){errors.push("No Date specified");}
    if (errors.length)
    {
        res.status(400).json({"error":errors.join(",")});
        console.log("Error adding album.")
        return;
    }

    var data = {
        name: req.body.name,
        user_id: Number(req.body.user_id),
        date: req.body.creation_date
    }

    var sql ='INSERT INTO Albums (Name, User_id, Creation_date) VALUES (?,?,?)'
    var params =[data.name, data.user_id, data.date]
    
    db.run(sql, params, function (err, result) {
        if (err)
        {
            res.status(400).json({"error": err.message})
            console.log("Error adding album.")
            return;
        }
        console.log("New album added.")
        res.json({
            "message": "success",
            "data": data,
            "id" : this.lastID
        })
    });
});

// Add new photo
app.post("/api/picture/", (req, res, next) => {
    var errors=[]
    if (!req.body.album_id){errors.push("No Album_id specified");}
    if (!req.body.creation_date){errors.push("No Date specified");}
    if (!req.body.location){errors.push("No Location specified");}
    if (errors.length)
    {
        res.status(400).json({"error":errors.join(",")});
        console.log("Error adding photo.")
        return;
    }

    var data = {
        name: req.body.name,
        album_id: Number(req.body.album_id),
        date: req.body.creation_date,
        location: req.body.location
    }

    var sql ='INSERT INTO Pictures (Name, Location, Creation_date, Album_id) VALUES (?,?,?,?)'
    var params =[data.name, data.location, data.date, data.album_id]
    
    db.run(sql, params, function (err, result) {
        if (err)
        {
            res.status(400).json({"error": err.message})
            console.log("Error adding photo.")
            return;
        }
        console.log("New photo added.")
        res.json({
            "message": "success",
            "data": data,
            "id" : this.lastID
        })
    });
});

// Add new tag
app.post("/api/tag/", (req, res, next) => {
    var errors=[]
    if (!req.body.picture_id){errors.push("No Picture_id specified");}
    if (!req.body.user_id){errors.push("No User_id specified");}
    if (errors.length)
    {
        res.status(400).json({"error":errors.join(",")});
        console.log("Error adding tag.")
        return;
    }

    var data = {
        user_id: Number(req.body.user_id),
        picture_id: Number(req.body.picture_id)
    }

    var sql ='INSERT INTO Tags (Picture_id, User_id) VALUES (?,?)'
    var params =[data.picture_id, data.user_id]
    
    db.run(sql, params, function (err, result) {
        if (err)
        {
            res.status(400).json({"error": err.message})
            console.log("Error adding tag.")
            return;
        }
        console.log("New tag added.")
        res.json({
            "message": "success",
            "data": data,
            "id" : this.lastID
        })
    });
});

// Update user data
app.patch("/api/user/:id", (req, res, next) => {
    var data = {
        password :md5(req.body.password)}

    db.run(
        `UPDATE Users set
           password = COALESCE(?,password) 
           WHERE id = ?`,
        [data.password, req.params.id],
        function (err, result)
        {
            if (err){
                res.status(400).json({"error": res.message})
                console.log("Error update user.")
                return;
            }
            console.log("User update.")
            res.json({
                message: "success",
                data: data,
                changes: this.changes
            })
    });
});

// Update album data
app.patch("/api/album/:id", (req, res, next) => {
    var data = {name: req.body.name}

    db.run(
        `UPDATE Albums set 
           name = COALESCE(?,name) 
           WHERE User_id = ?`,
        [data.name, req.params.id],
        function (err, result)
        {
            if (err){
                res.status(400).json({"error": res.message})
                console.log("Error update album.")
                return;
            }
            console.log("Album update.")
            res.json({
                message: "success",
                data: data,
                changes: this.changes
            })
    });
});

// Update photo data
app.patch("/api/picture/:id", (req, res, next) => {
    var data = {
        name: req.body.name,
        location: req.body.location}

    db.run(
        `UPDATE Pictures set 
           name = COALESCE(?,name),
           location = COALESCE(?,location)
           WHERE Album_id = ?`,
        [data.name, data.location, req.params.id],
        function (err, result)
        {
            if (err){
                res.status(400).json({"error": res.message})
                console.log("Error update photo.")
                return;
            }
            console.log("Photo update.")
            res.json({
                message: "success",
                data: data,
                changes: this.changes
            })
    });
});

// Delete user by id
app.delete("/api/user/:id", (req, res, next) => {
    var sql = 'DELETE FROM Users WHERE id = ' +req.params.id
    var params = []

    db.run(sql, params, function (err, result) {
        if (err)
        {
            res.status(400).json({"error": err.message})
            console.log("Error delete user.")
            return;
        }
        console.log("Delete user.")
        res.json({"message":"deleted", changes: this.changes});
    });
});

// Delete album by name
app.delete("/api/album/:name", (req, res, next) => {
    var sql = 'DELETE FROM Albums WHERE Name = "' + req.params.name + '"'
    var params = []

    db.run(sql, params, function (err) {
        if (err)
        {
            res.status(400).json({"error": err.message})
            console.log("Error delete album.")
            return;
        }

        console.log("Delete album.")
        res.json({"message":"deleted", changes: this.changes});
    });
});

// Delete photo by album id + user name
app.delete("/api/picture/:album_id/:name", (req, res, next) => {
    var sql = 'DELETE FROM Pictures WHERE Name = "' + req.params.name + '" AND Album_id = ' + req.params.album_id
    var params = []

    db.run(sql, params, function (err) {
        if (err)
        {
            res.status(400).json({"error": err.message})
            console.log("Error delete photo.")
            return;
        }

        console.log("Delete photo.")
        res.json({"message":"deleted", changes: this.changes});
    });
});

// Delete tag by userid + picture id
app.delete("/api/tag/:user_id/:picture_id", (req, res, next) => {
    var sql = 'DELETE FROM Tags WHERE User_id = ' + req.params.user_id + ' AND Picture_id = ' + req.params.picture_id
    var params = []

    db.run(sql, params, function (err) {
        if (err)
        {
            res.status(400).json({"error": err.message})
            console.log("Error delete tag.")
            return;
        }

        console.log("Delete tag.")
        res.json({"message":"deleted", changes: this.changes});
    });
});

// Check if pass is correct
app.post("/api/check/", (req, res, next) => {
    var errors=[]
    if (!req.body.password){errors.push("No password specified");}
    if (errors.length)
    {
        console.log("Error checking user.")
        res.status(400).json({"error":errors.join(",")});
        return;
    }

    var data = {
        name: req.body.username,
        password: md5(req.body.password)
    }

    var sql ='SELECT * FROM Users WHERE Name = ?'
    var params =[data.name]
    
    db.all(sql, params, function (err, result) {
        if (err)
        {
            res.status(400).json({"error": err.message})
            console.log("Error checking user.")
            return;
        }   

        console.log("User checked.")
        if (result.length == 0)
        {
            res.json({
                "message": "success",
                "data": "false"
            });
        }
        else if (result[0].Password == data.password)
        {
            res.json({
                "message": "success",
                "data": "true"
            });
        }
        else
        {
            res.json({
                "message": "success",
                "data": "false"
            });
        }
    });
});

// Default response for any other request
app.use(function(req, res){
    res.status(404);
});

File: GalleryGui/GalleryGui/GalleryApi.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Net.Http;
using Newtonsoft.Json.Linq;

namespace GalleryGui
{
    class GalleryApi
    {
        private string url = "http://127.0.0.1:3232/api/";

        #region Loggin/register things
        /*
         * Function check if user exist in database.
         */
        public bool userExist(string username)
        {
            // Create the get request msg
            string condition = url + "*/users/WHERE Name = '" + username + "'";

            // Get the response from the server.
            string rsponse;
            try
            {
                rsponse = getRequest(condition);
            }
            catch (Exception err)
            {
                throw err;
            }

            // Convert data to json.
            JObject json = JObject.Parse(rsponse);
            JArray data = (JArray) json.SelectToken("data"); // Take the only the "data" 

            // If the data is empty, username not exist.
            if (data == null)
                return false;
            if (data.Count == 0)
                return false;

            return true;
        }

        /*
         * The function create new user.
         */
        public void createUser(string username, string password)
        {
            string urlCreate = this.url + "user/";
            string data = "name=" + username + "&password=" + password;

            string response;
            try
            {
                response = postRequest(urlCreate, data);
            }
            catch (Exception err)
            {
                throw err;
            }

            // Convert data to json.
            JObject json = JObject.Parse(response);

            if (json.ContainsKey("error"))
                throw new Exception("Error while creating the account.");
        }

        /*
         * Function check if password is correct.
         */
        public bool tryLoggin(string username, string password)
        {
            string urlCreate = this.url + "check/";
            string data = "username=" + username + "&password=" + password;

            string response;
            try
            {
                response = postRequest(urlCreate, data);
            }
            catch (Exception err)
            {
                throw err;
            }

            // Convert data to json.
            JObject json = JObject.Parse(response);
            JValue checkData = (JValue)json.SelectToken("data");

            if ((string)checkData.Value == "true")
                return true;
            return false;
        }
        
        /*
         * The function update password
         */
        public void changePass(string userid, string password)
        {
            string urlCreate = this.url + "user/" + userid;
            string data = "password=" + password;

            string response;
            try
            {
                response = patchRequest(urlCreate, data);
            }
            catch (Exception err)
            {
                throw err;
            }

            // Convert data to json.
            JObject json = JObject.Parse(response);

            if (json.ContainsKey("error"))
                throw new Exception("Error while changing password");

            else if (json.Count == 0)
                throw new Exception("Error while changing password");
        }
        #endregion

        #region Get things
        /*
         * The function reture the user id
         */
        public string getUserId(string username)
        {
            string userId;

            // Create the get request msg
            string condition = url + "*/users/WHERE Name = '" + username + "'";

            // Get the response from the server.
            string rsponse;
            try
            {
                rsponse = getRequest(condition);
            }
            catch (Exception err)
            {
                throw err;
            }

            // Convert data to json.
            JObject json = JObject.Parse(rsponse);
            var data = json.SelectToken("data"); // Take the only the "data"

            userId = data.First.SelectToken("ID").ToString();
            return userId;
        }

        /*
         * The function return list of all users 
         */
        public List<string> getAllUsers()
        {
            List<string> usersList = new List<string>();

            // Create the get request msg
            string condition = url + "*/users/null";

            // Get the response from the server.
            string rsponse;
            try
            {
                rsponse = getRequest(condition);
            }
            catch (Exception err)
            {
                throw err;
            }

            // Convert data to json.
            JObject json = JObject.Parse(rsponse);
            JArray data = (JArray)json.SelectToken("data"); // Take the only the "data"

            foreach (var field in data)
            {
                string name = field.SelectToken("Name").ToString();
                usersList.Add(name);
            }

            return usersList;
        }

        /*
         * The function return all the albums
         */
        public List<Album> getAllAlbums()
        {
            List<Album> albumsList = new List<Album>();

            // Create the get request msg
            string condition = url + "*/Albums/null";

            // Get the response from the server.
            string rsponse;
            try
            {
                rsponse = getRequest(condition);
            }
            catch (Exception err)
            {
                throw err;
            }

            // Convert data to json.
            JObject json = JObject.Parse(rsponse);
            JArray data = (JArray)json.SelectToken("data"); // Take the only the "data"

            foreach (var field in data)
            {
                Album Album = new Album();
                Album.name = field.SelectToken("Name").ToString();
                Album.ID = int.Parse(field.SelectToken("ID").ToString());
                Album.date = field.SelectToken("Creation_date").ToString();
                Album.ownerID = int.Parse(field.SelectToken("User_id").ToString());

                albumsList.Add(Album);
            }

            return albumsList;
        }

        /* 
         * The function return all the photos in album
         */
        public List<Img> getAlbumPhotos(Album album)
        {
            List<Img> photosList = new List<Img>();

            // Create the get request msg
            string condition = url + "*/Pictures/WHERE Album_id = " + album.ID;

            // Get the response from the server.
            string rsponse;
            try
            {
                rsponse = getRequest(condition);
            }
            catch (Exception err)
            {
                throw err;
            }

            // Convert data to json.
            JObject json = JObject.Parse(rsponse);
            JArray data = (JArray)json.SelectToken("data"); // Take the only the "data"

            foreach (var field in data)
            {
                Img photo = new Img();
                photo.name = field.SelectToken("Name").ToString();
                photo.ID = int.Parse(field.SelectToken("ID").ToString());
                photo.date = field.SelectToken("Creation_date").ToString();
                photo.Path = field.SelectToken("Location").ToString();
                photo.albumID = int.Parse(field.SelectToken("Album_id").ToString());

                photosList.Add(photo);
            }

            return photosList;
        }

        /*
         * The function return an owner of photo
         */
        public int getPhotoOwner(Img photo)
        {
            int ownerID = 0;

            // Create the get request msg
            string condition = url + "Albums.User_id as owner/Albums/JOIN Pictures ON Pictures.Album_id = Albums.ID WHERE Pictures.ID = " + photo.ID;

            // Get the response from the server.
            string rsponse;
            try
            {
                rsponse = getRequest(condition);
            }
            catch (Exception err)
            {
                throw err;
            }

            // Convert data to json.
            JObject json = JObject.Parse(rsponse);
            JArray data = (JArray)json.SelectToken("data"); // Take the only the "data"

            if (data != null)
            {
                ownerID = int.Parse(data.First.SelectToken("owner").ToString());
            }

            return ownerID;
        }

        /*
         * The function return the users that are tagged in picture.
         */
        public List<string> getTaggedUsers(Img photo)
        {
            List<string> usersList = new List<string>();

            // Create the get request msg
            string condition = url + "Users.Name/Users/JOIN Tags ON Users.ID = Tags.User_id WHERE Tags.Picture_id = " + photo.ID;

            // Get the response from the server.
            string rsponse;
            try
            {
                rsponse = getRequest(condition);
            }
            catch (Exception err)
            {
                throw err;
            }

            // Convert data to json.
            JObject json = JObject.Parse(rsponse);
            JArray data = (JArray)json.SelectToken("data"); // Take the only the "data"

            foreach (var field in data)
            {
                string name = field.SelectToken("Name").ToString();
                usersList.Add(name);
            }

            return usersList;
        }

        /*
         * The function return the amount of tags in album
         */
        public int albumTagsAmount(Album album)
        {
            int amount = 0;

            // Create the get request msg
            string condition = url + "count(Tags.ID) as amount/Tags/JOIN Pictures ON Tags.Picture_id = Pictures.ID JOIN Albums ON Pictures.Album_id = Albums.ID WHERE Albums.id = " + album.ID;

            // Get the response from the server.
            string rsponse;
            try
            {
                rsponse = getRequest(condition);
            }
            catch (Exception err)
            {
                throw err;
            }

            // Convert data to json.
            JObject json = JObject.Parse(rsponse);
            JArray data = (JArray)json.SelectToken("data"); // Take the only the "data"

            if (data != null)
            {
                amount = int.Parse(data.First.SelectToken("amount").ToString());
            }

            return amount;
        }

        /*
         * The function return the location of the last 4 photos.
         * (HomePage)
         */
        public List<Img> getLastPhotos()
        {
            List<Img> photosList = new List<Img>();

            // Create the get request msg
            string condition = url + "*/Pictures/LIMIT 4";

            // Get the response from the server.
            string rsponse;
            try
            {
                rsponse = getRequest(condition);
            }
            catch (Exception err)
            {
                throw err;
            }

            // Convert data to json.
            JObject json = JObject.Parse(rsponse);
            JArray data = (JArray)json.SelectToken("data"); // Take the only the "data"

            foreach (var field in data)
            {
                Img photo = new Img();
                photo.name = field.SelectToken("Name").ToString();
                photo.ID = int.Parse(field.SelectToken("ID").ToString());
                photo.date = field.SelectToken("Creation_date").ToString();
                photo.Path = field.SelectToken("Location").ToString();
                photo.albumID = int.Parse(field.SelectToken("Album_id").ToString());

                photosList.Add(photo);
            }

            return photosList;
        }

        /*
         * The function return album data
         */
        public Album openAlbum(string albumName)
        {
            Album album = new Album();

            // Create the get request msg
            string condition = url + "*/Albums/WHERE Name = '" + albumName + "'";

            // Get the response from the server.
            string rsponse;
            try
            {
                rsponse = getRequest(condition);
            }
            catch (Exception err)
            {
                throw err;
            }

            // Convert data to json.
            JObject json = JObject.Parse(rsponse);
            JArray data = (JArray)json.SelectToken("data"); // Take the only the "data"

            foreach (var field in data)
            {
                album.name = field.SelectToken("Name").ToString();
                album.ID = int.Parse(field.SelectToken("ID").ToString());
                album.date = field.SelectToken("Creation_date").ToString();
                album.ownerID = int.Parse(field.SelectToken("User_id").ToString());
            }

            return album;
        }

        /*
         * The function check if the name exist in the album
         */
        public bool CheckNameInAlbum(Album album, string nameToCheack)
        {
            List<Img> imgList = getAlbumPhotos(album);

            foreach(Img photo in imgList)
            {
                if (photo.name == nameToCheack)
                    return true;
            }

            return false;
        }

        /*
         * The function check if the album name is taken
         */
        public bool CheckAlbumName(string albumName)
        {
            List<Album> albums = getAllAlbums();

            foreach(Album album in albums)
            {
                if (album.name == albumName)
                    return true;
            }

            return false;
        }

        /*
         * The function return all the picture of user
         */
        public List<Img> getAllPhotosOf(string userid)
        {
            List<Img> userPhotos = new List<Img>();
            List<Album> userAlbums = getAllAlbumsOf(userid);

            foreach(Album album in userAlbums)
            {
                List<Img> newImgs = getAlbumPhotos(album);
                foreach (Img img in newImgs)
                {
                    userPhotos.Add(img);
                }
            }

            return userPhotos;
        }

        /*
         * The function return all the albums of user
         */
        public List<Album> getAllAlbumsOf(string userid)
        {
            List<Album> allAlbums = getAllAlbums();
            List<Album> userAlbums = new List<Album>();

            foreach(Album album in allAlbums)
            {
                if (album.ownerID.ToString() == userid)
                {
                    userAlbums.Add(album);
                }
            }

            return userAlbums;
        }

        /*
         * The function return the tags of user
         */
        public int getTagsCount(string userid)
        {
            int count = 0;

            // Create the get request msg
            string condition = url + "count(id) as count/Tags/WHERE user_id = " + userid;

            // Get the response from the server.
            string rsponse;
            try
            {
                rsponse = getRequest(condition);
            }
            catch (Exception err)
            {
                throw err;
            }

            // Convert data to json.
            JObject json = JObject.Parse(rsponse);
            JArray data = (JArray)json.SelectToken("data");

            if (data != null)
                count = int.Parse(data.First.SelectToken("count").ToString());

            return count;
        }

        #endregion

        #region Create things
        /*
         * The function create new album
         */
        public void createAlbum(Album newAlbum)
        {
            string urlCreate = this.url + "album/";
            string data = "name=" + newAlbum.name + "&creation_date=" + newAlbum.date + "&user_id=" + newAlbum.ownerID;

            string response;
            try
            {
                response = postRequest(urlCreate, data);
            }
            catch (Exception err)
            {
                throw err;
            }

            // Convert data to json.
            JObject json = JObject.Parse(response);

            if (json.ContainsKey("error"))
                throw new Exception("Error while creating the album.");
        }

        /*
         * The function create new photo
         */
        public void createImage(Img newPhoto)
        {
            string urlCreate = this.url + "picture/";
            string data = "name=" + newPhoto.name + "&creation_date=" + newPhoto.date + "&location=" + newPhoto.Path + "&album_id=" + newPhoto.albumID;

            string response;
            try
            {
                response = postRequest(urlCreate, data);
            }
            catch (Exception err)
            {
                throw err;
            }

            // Convert data to json.
            JObject json = JObject.Parse(response);

            if (json.ContainsKey("error"))
                throw new Exception("Error while creating the photo.");
        }

        /*
         * The function create new tag
         */
        public void addTag(string imgID, string userID)
        {
            string urlCreate = this.url + "tag/";
            string data = "picture_id=" + imgID + "&user_id=" + userID;

            string response;
            try
            {
                response = postRequest(urlCreate, data);
            }
            catch (Exception err)
            {
                throw err;
            }

            // Convert data to json.
            JObject json = JObject.Parse(response);

            if (json.ContainsKey("error"))
                throw new Exception("Error while creating tag.");
        }
        #endregion

        #region delete things.
        /*
         * The function delete a user.
         */
        public void deleteUser(string userId)
        {
            // Delete all users albums
            List<Album> userAlbums = getAllAlbumsOf(userId);
            foreach(Album album in userAlbums)
            {
                deleteAlbum(album);
            }

            string deleteURL = url + "user/" + userId;

            string response;
            try
            {
                response = deleteRequest(deleteURL);
            }
            catch (Exception err)
            {
                throw err;
            }

            // Convert data to json.
            JObject json = JObject.Parse(response);

            if (json.ContainsKey("error"))
                throw new Exception("Error deleting account.");
        }

        /*
         * The function delete an album
         */
        public void deleteAlbum(Album album)
        {
            List<Img> photosList = getAlbumPhotos(album);

            // Delete all photos in album
            foreach(Img photo in photosList)
            {
                deletePhoto(photo);
            }

            string deleteURL = url + "album/" + album.name;

            string response;
            try
            {
                response = deleteRequest(deleteURL);
            }
            catch (Exception err)
            {
                throw err;
            }

            // Convert data to json.
            JObject json = JObject.Parse(response);

            if (json.ContainsKey("error"))
                throw new Exception("Error deleting album.");
        }

        /*
         * The function delete a photo
         */
        public void deletePhoto(Img photo)
        {
            List<string> usersTaged = getTaggedUsers(photo);

            // Delete all tags in picture.
            foreach (string userName in usersTaged)
            {
                string userID = getUserId(userName);
                deleteTag(photo.ID.ToString(), userID);
            }

            string deleteURL = url + "picture/" + photo.albumID + "/" + photo.name;

            string response;
            try
            {
                response = deleteRequest(deleteURL);
            }
            catch (Exception err)
            {
                throw err;
            }

            // Convert data to json.
            JObject json = JObject.Parse(response);

            if (json.ContainsKey("error"))
                throw new Exception("Error deleting photo.");
        }

        /* 
         * The function delete a tag
         */
        public void deleteTag(string photoID, string userID)
        {
            string deleteURL = url + "tag/" + userID + "/" + photoID;

            string response;
            try
            {
                response = deleteRequest(deleteURL);
            }
            catch (Exception err)
            {
                throw err;
            }

            // Convert data to json.
            JObject json = JObject.Parse(response);

            if (json.ContainsKey("error"))
                throw new Exception("Error deleting tag.");
        }
        #endregion

        #region Post,Get,Delete,Patch Requests hendlers.

        /*
        * Function get the data from the PostRquest and conver it to string
        */
        private string postRequest(string url, string data)
        {
            List<KeyValuePair<string, string>> postData = new List<KeyValuePair<string, string>>();
            List<string> dataSplit = data.Split('&').ToList();

            foreach (string pair in dataSplit)
            {
                string[] pairArr = pair.Split('=').ToArray<string>();
                postData.Add(new KeyValuePair<string, string>(pairArr[0], pairArr[1]));
            }

            string result;
            try
            {
                result = Task.Run(async () => await PostRequest(url, postData)).Result;
            }
            catch
            {
                throw new Exception("Something wrong with the server\n Please check that the server is on.");
            }

            return result;
        }

        /*
         * Function get the data from the GetRquest and conver it to string
         */
        private string getRequest(string data)
        {
            string result;
            try
            {
                result = Task.Run(async () => await GetRequest(data)).Result;
            }
            catch
            {
                throw new Exception("Something wrong with the server\n Please check that the server is on.");
            }

            return result;
        }

        /*
         * Function get the data from the PatchRquest and conver it to string
         */
        private string patchRequest(string url, string data)
        {
            List<KeyValuePair<string, string>> postData = new List<KeyValuePair<string, string>>();
            List<string> dataSplit = data.Split('&').ToList();

            foreach (string pair in dataSplit)
            {
                string[] pairArr = pair.Split('=').ToArray<string>();
                postData.Add(new KeyValuePair<string, string>(pairArr[0], pairArr[1]));
            }

            string result;
            try
            {
                result = Task.Run(async () => await PatchRequest(url, postData)).Result;
            }
            catch
            {
                throw new Exception("Something wrong with the server\n Please check that the server is on.");
            }

            return result;
        }

        /*
         * Function get the data from the DeleteRquest and conver it to string
         */
        private string deleteRequest(string url)
        {
            string result;
            try
            {
                result = Task.Run(async () => await DeleteRequest(url)).Result;
            }
            catch
            {
                throw new Exception("Something wrong with the server\n Please check that the server is on.");
            }

            return result;
        }

        /*
         * Function handle with the Get request
         */
        private async static Task<string> GetRequest(string url)
        {
            using (HttpClient client = new HttpClient())
            {
                using (HttpResponseMessage response = await client.GetAsync(url))
                {
                    using (HttpContent data = response.Content)
                    {
                        string content = await data.ReadAsStringAsync();
                        return content;
                    }
                }
            }
        }

        /*
         * Functionn handle with the Post request
         */
        private async static Task<string> PostRequest(string url, List<KeyValuePair<string, string>> postData)
        {

            IEnumerable<KeyValuePair<string, string>> queries = postData;
            HttpContent postDataContent = new FormUrlEncodedContent(queries);

            using (HttpClient client = new HttpClient())
            {
                using (HttpResponseMessage response = await client.PostAsync(url, postDataContent))
                {
                    using (HttpContent data = response.Content)
                    {
                        string content = await data.ReadAsStringAsync();
                        return content;
                    }
                }
            }
        }

        /*
         * Functionn handle with the Patch request
         */
        private async static Task<string> PatchRequest(string url, List<KeyValuePair<string, string>> postData)
        {
            IEnumerable<KeyValuePair<string, string>> queries = postData;
            HttpContent postDataContent = new FormUrlEncodedContent(queries);
            using (HttpClient client = new HttpClient())
            {
                using (HttpResponseMessage response = await Patch.PatchAsync(client, url, postDataContent))
                {
                    using (HttpContent data = response.Content)
                    {
                        string content = await data.ReadAsStringAsync();
                        return content;
                    }
                }
            }
        }

        /*
        * Functionn handle with the Delete request
        */
        private async static Task<string> DeleteRequest(string url)
        {
            using (HttpClient client = new HttpClient())
            {
                using (HttpResponseMessage response = await client.DeleteAsync(url))
                {
                    using (HttpContent data = response.Content)
                    {
                        string content = await data.ReadAsStringAsync();
                        return content;
                    }
                }
            }
        }
        #endregion
    }

    static class Patch
    {
        public static async Task<HttpResponseMessage> PatchAsync(this HttpClient client, string requestUri, HttpContent iContent)
        {
            var method = new HttpMethod("PATCH");
            var request = new HttpRequestMessage(method, requestUri)
            {
                Content = iContent
            };

            HttpResponseMessage response = new HttpResponseMessage();
            try
            {
                response = await client.SendAsync(request);
            }
            catch (TaskCanceledException e)
            {
                throw e;
            }

            return response;
        }

    }

    public class Img
    {
        public string Path { get; set; }
        public string name { get; set; }
        public string date { get; set; }
        public int albumID { get; set; }
        public int ID { get; set; }
    }

    public class Album
    { 
        public int ID { get; set; }
        public string name { get; set; }
        public string date { get; set; }
        public int ownerID { get; set; }
    }
}
